L85-core-binder-transport.md

```diff
@@ -0,0 +1,181 @@
+when `GPR_ANDROID` is defined)
+
+### Server side API
+
+Service implementation will be an Android [bound service] in Java, and it should return
+a binder object created by our transport implementation to the client.
+
+We propose to introduce a new API that returns credentials for binder server:
+
+```Cpp
+std::shared_ptr<ServerCredentials> BinderServerCredentials(std::shared_ptr<grpc::binder::SecurityPolicy> security_policy);
+```
+
+The following snippet sets up gRPC server listening for incoming binder
+transactions, and only allows clients that signed by the same key to connect:
+
+```Cpp
+grpc::ServerBuilder server_builder;
+server_builder.AddListeningPort(
+    "binder://example",
+    grpc::BinderServerCredentials(grpc::binder::SameSignatureSecurityPolicy()));
+```
+
+The first argument specifies a "binder port". A binder port is not a real port,
+and is identified with a custom binder URI scheme `binder://`.
+
+After `server_builder.BuildAndStart()` is called, an endpoint binder will be
+created internally in the transport.
+
+To let the client connect to the server, the service needs to return the
+endpoint binder. To get the endpoint binder, we provide a Java class `GrpcCppServerBuilder` and a static Java method
+that can be used like the following:
+
+```java
+GrpcCppServerBuilder::GetEndpointBinder("example")
+```
+
+The string parameter corresponds to the first argument passed in
+`grpc::ServerBuilder::AddListeningPort`.
+
+Note that the native library containing our transport implementation must
+have been loaded by user before using this helper class.
+
+Under the hood we will use a shared static variable to store the mapping between
+service name and endpoint binder.
+
+## Notes
+
+### Why asynchronous callback is used in client side API
+
+Due to Android API's design, Android service might not call
+`onServiceConnected` before current activity ends.
+
+User should not block the (android) Activity to wait for the callback to be
+called. Instead, user should unblock the thread and wait Android to notify that
+the bind is successful, in an asynchronous manner.
+
+We made the design decision to not return a (pending) channel object immediately
+because typically user will want to be notified when the channel become usable, and
+we want to prevent user from using the (not connected) channel immediately and
+wait for RPC call to success and causing deadlock.
+
+### Parcelable objects
+
+The protocol and Java implementation allows passing Parcelable objects. We
+currently don't have a plan to support it in C++. When we see a parcelable
+object in `Parcel`, we will print an error message and fail the RPC call.
+
+### Error handling
+
+We will mark the API as experimental for now and experiment it in real world use
+cases. After that we will integrate proper error handling into our interface.
+
+### Android API level requirement
+
+For Java interoperability, we will require the device to have API level 32
```

Comment:
What features require API level 32? It seems unfortunate that this would rule out use of BinderChannel on all existing Android devices until receiving an OS update. Is it possible to target a lower API level?

Reply:
[A new API for disabling header](https://cs.android.com/android/platform/superproject/+/master:frameworks/native/libs/binder/ndk/include_ndk/android/binder_ibinder.h;l=234?q=AIBinder_Class_disableInterfaceTokenHeader&ss=android%2Fplatform%2Fsuperproject)

> Is it possible to target a lower API level?

For C++ Server <--> C++ Client scenario it is possible to target API level 31